torchdata/dataloader2/reading_service.py

```diff
@@ -216,11 +224,20 @@ def initialize(self, datapipe: DataPipe) -> DataPipe:
             self.datapipes.append(local_datapipe)
 
         self.combined_datapipes = _IterateQueueDataPipes(self.datapipes)
+        if self.prefetch_mainloop > 0:
+            self.combined_datapipes = self.combined_datapipes.prefetch(self.prefetch_mainloop)
         return self.combined_datapipes  # type: ignore[return-value]
 
     def initialize_iteration(self) -> None:
         if self.combined_datapipes is not None:
-            self.combined_datapipes.reset_epoch()
+            if self.prefetch_mainloop > 0:
+                # Stop prefetching first
+                self.combined_datapipes.reset()
```

Comment:
Since `reset` will be called in the begining of the `__iter__` as well, I think it's better to add a different API to stop prefetching and called here.

And, `reset` can be the last resort to stop thread.

Reply:
TBH, we should have a graph function to stop `prefetcher` from the end of the pipeline to the beginning especially when multiple `prefetcher`s are presented.

Reply:
I also found them annoyingly duplicating and ideally we need something like GraphStop operation to freeze all parallel work. It would be extremely useful for snapshotting. But it is out of scope for this PR (Anyway I will post separate issue to capture demand)

Reply:
SGTM! There are a few special DataPipes need to be handled specially with this `GraphStop`:
- `FullSync`
- `Prefetcher`